src/DotNet.Status.Web/Controllers/AzurePipelinesController.cs

```diff
@@ -361,6 +400,125 @@ private async Task<string> BuildChangesMessage(Build build)
+            bool monitorHasTags)
+        {
+            string repoChangesMessage = "";
+
+            if(TryGetValidatingBarIds(build.Tags, out List<int> validatingBarIds))
+            {
+                // We are interested in looking at the diff in the repo of interest, so:
+                // 1) Figure out what repo the bar id corresponds to
+                // 2) Find the last successful build of this pipeline that corresponds to that repo and get its ValidatingBarIds tags
+                // 3) Get the commit sha of the two bar ids, and get the commit diff between those two builds
+
+                // Get the previous day's builds of this definition
+                DateTimeOffset minTime = DateTimeOffset.Parse(build.QueueTime).AddDays(-1);
+                Build[] previousBuilds = await Client.ListBuilds(
+                    build.Project.Name, 
+                    CancellationToken.None, 
+                    minTime: minTime, 
+                    maxTime: DateTimeOffset.Parse(build.QueueTime), 
+                    limit: default,
+                    definitions: build.Definition.Id.ToString());
+                List<Build> matchingBuilds = new List<Build>();
+
+                // For monitors that have tags, only get the builds with matching tags. For all other monitors, use the full list
+                if (monitorHasTags)
+                {
+                    _logger.LogInformation($"Choosing builds that have the same tags as the monitor.");
+                    matchingBuilds = previousBuilds
+                        .Where(b => !IsIgnoredReason(b))
+                        .Where(b => 
+                            monitor.Tags.Intersect(b.Tags).SequenceEqual(monitor.Tags) 
+                            && DateTimeOffset.Parse(b.StartTime) < DateTimeOffset.Parse(build.StartTime))
+                        .OrderByDescending(b => b.StartTime)
+                        .ToList();
+                }
+                else
+                {
+                    _logger.LogInformation($"Ignoring tags, as monitor doesn't specify them.");
+                    matchingBuilds = previousBuilds
+                        .Where(b => !IsIgnoredReason(b))
+                        .OrderByDescending(b => b.StartTime)
+                        .ToList();
+                }
+
+                // If there are any builds from the day before that were successful, use that, otherwise, just get the most recent previous build
+                Build compareBuild = matchingBuilds.FirstOrDefault(b => b.Result == "succeeded");
+
+                if (compareBuild == null)
+                {
+                    compareBuild = matchingBuilds.FirstOrDefault();
+                }
+
+                // Get the build information for the build that was being validated, so that we can get the change history
+                MaestroBuild barBuild = await GetBuildAsync(validatingBarIds.First());
```

Comment:
I REALLY don't like the added maestro dependency here.  I assume this is necessary, but it's adding a lot of complication to this relatively simple webhook. :-(

Reply:
Let me look at how I'm using this. I think we're just using it so we can get the info about the azdo run that did the publishing, and we had decided to save away the bar id. Maybe we can have the official builds mark down the azdo id and then we can use some azdo apis to get the commit history.

Reply:
Created https://github.com/dotnet/arcade-validation/pull/2895 so that the repo that is using this will be able to just supply the azdo build url, and I will change this to parse the url to get the azdo build id/project.

Reply:
Parsing URL's is rather fragile.  Can we just tag the project/build directly rather than the whole URL?

Reply:
We are going to have to parse it somewhere. Darc's get-asset/get-build only gives you the build link, not the, say, build id. I could either parse it here (where we could make it consistent for anyone who wanted to use this functionality) or parse it everywhere we need to convert a bar id to an azdo build id/project. I'd rather parse it here, personally (1 place to maintain the code).

Reply:
Ugh.  We really shouldn't be parsing build URL's.  We just ran into a huge problem in the build analysis because of it.  They are not a reliable means of transmitting information, and I'm sad that its tendrils are reaching out further.

Reply:
If we have to we have to... just try to make sure we behave reasonably if we fail to parse the URL.